odl/deform/linearized_deform.py

```diff
@@ -0,0 +1,546 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Operators of linearized deformations.
+"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+from odl.operator.operator import Operator
+from odl import ProductSpace, DiscreteLp, Gradient, PointwiseInner, Divergence
+import numpy as np
+
+
+__all__ = ('LinDeformFixedTempl', 'LinDeformFixedTemplDeriv',
+           'LinDeformFixedTemplDerivAdj', 'LinDeformFixedDisp',
+           'LinDeformFixedDispAdj', 'linear_deform', 'deform_grad')
+
+
+def linear_deform(template, displacement):
+    """Linearized deformation of a template with a displacement field.
+
+    The template ``I`` is deformed by a displacement field ``v`` as:
+
+        (I, v) --> I(Id + v)
+
+    where ``Id`` is the identity mapping:
+
+        y --> y
+
+    and the vector field ``v`` is a displacement field:
+
+        y --> v(y)
+
+    Here, the ``y`` is an element in the domain.
+
+    Parameters
+    ----------
+    template : `DiscreteLp` element
+        Template to be deformed by a displacement field.
+    displacement : `ProductSpace` element
+        The vector field (displacement field) used in the linearized
+        deformation.
+
+    Returns
+    -------
+    deformed_template : `numpy.ndarray`
+        Deformed template as an numpy array.
+
+    Examples
+    --------
+    >>> import odl
+    >>> space = odl.uniform_discr(0, 1, 5)
+    >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+    >>> linear_deform(template, displacement_field)
+    array([ 0.,  0.,  1.,  1.,  0.])
+    """
+    image_pts = template.space.grid.points()
+    image_pts += np.asarray(displacement).T
```

Comment:
This code will be extremely slow. Simple test here:
```python
>>> discr = odl.uniform_discr([-1, -1], [1, 1], (1000, 1000))
>>> pspace = odl.ProductSpace(discr, discr.ndim)
>>> v = pspace.one()
>>>
>>> def f1(v):
...     arr = np.empty((len(v),) + v[0].shape, dtype=v[0].dtype)
...     for i, vi in enumerate(v):
...         arr[i, ...] = vi
...     return arr
>>>
>>> def f2(v):
...     return np.asarray(v)
>>>
>>> %timeit f1(v)
The slowest run took 5.68 times longer than the fastest. This could mean that an intermediate result is being cached.
100 loops, best of 3: 1.71 ms per loop

>>> %timeit f2(v)
1 loop, best of 3: 5.51 s per loop
```
So we need to manually update the image points (which will be faster and lighter on memory since we don't create a copy of `displacement`):
```python
>>> for i, vi in enumerate(displacement):
...     image_pts[:, i] += vi.asarray().ravel()
```